docs/unidling.md

```diff
@@ -0,0 +1,185 @@
+10. The kube-proxy receives a watch event with the new endpoint information and updates its routing
+    table
+11. The kube-proxy services the request to the new endpoint
+
+### Thrash prevention
+
+The unidler must be able to correctly handle bursts of `WantPods` events for the same service.
+This implies that the unidler has an internal representation of which services have recently been
+unidled and that the unidle process itself has checks to ensure that only the minimum number of
+pods to unidle a service are created.  Concerns about scaling upon unidle should be handled
+exclusively by the autoscaler.
+
+### Example: multiple kube-proxies receive requests for idled service
+
+Thrash prevention becomes relevant when multiple kube-proxies receive requests for an idled service
+at approximately the same time.
+
+1.  The kube-proxy on node N1...Nx receive requests for S
+2.  The kube-proxy socket handlers create `WantPods` events and wait X seconds for new
+    endpoints for S to be created
+3.  The unidler receives the first watch event for the `WantPods` event
+4.  The unidler resolves the S to some replication controller R
+5.  The unidler resizes R to *n* >= 1
+6.  The unidler receives subsequent `WantPods` events from other nodes and ignores them since
+    there is already a replication controller that resolves to S with *n* >= 1
+7.  The replication controller manager schedules a new pod P for R
+8.  The scheduler determines placement and creates a `Binding` for P
+9.  The kubelet on the appropriate node processes the `Binding` and runs P
+10. The endpoints controller determines that S has a new endpoint and updates S's endpoints
+11. The kube-proxies receive watch events with the new endpoint information and updates their routing
+    tables
+12. The kube-proxies service the requests to the new endpoint
+
+## Proposed Design
+
+### Modifications to kube-proxy
+
+The kube-proxy must be modified create events when idled services are requested and make an attempt
+to wait until endpoints are available.  Currently, the proxy delegates across a `LoadBalancer`
+interface to get the endpoint to proxy a connection to.  It makes sense to detect idled services
+inside this interface boundary.
+
+Currently, there is only one implementation of the `LoadBalancer` interface but we do not want to
+tie unidle functionality to that implementation.  There should be a new internal component of the
+proxier that can be used by any `LoadBalancer` implementation that will handle creating a
+`WantPods` event and handle waiting until a timeout for destination pods to become available.  The
+fields of the `WantPods` event will be set as described above.
+
+### The `UnidlerController`
+
+The `UnidlerController` will be added to `pkg/unidle`.  It will use the following algorithm to
+resolve a service S to replication controllers for unidling:
+
+1.  If there are any replication controllers in S's namespace with the annotation
```

Comment:
It's on the replication controller because we expect replication controllers to change more often than services.  I'm open to arguments for why that's bad, of course.